### ISSUE TITLE
Handle multiple PNGs at once

### ISSUE THREAD

**Issue by trueNAHO** [redwarp/optimize-png-hooks#1]
Running [oxipng](https://github.com/shssoichiro/oxipng) against a whole repository is almost instant. [This](https://github.com/redwarp/optimize-png-hooks) hook however, seems to struggle (at least on my end). Commiting every PNG one by one works just fine. Commiting multiple at once, seems to take forever. I waited for 30min and it was still not done. Doing the same one by one was done in a second.

Here a simple reproducible environment:

```shell
git clone 'https://github.com/Se7endAY/grub2-theme-vimix.git'
cd grub2-theme-vimix/
echo "repos:
  - repo: https://github.com/redwarp/optimize-png-hooks
    rev: v1.2.0
    hooks:
      - id: optimize-png" > .pre-commit-config.yaml
pre-commit install
mv Vimix/ src
```

Now commit PNG files one by one:

```shell
find src/ -maxdepth 1 -type f -name "*.png" -print0 | xargs -0 -I {} bash -c "git add {} && git commit -m 'add {}'; git add {} && git commit -m 'add {}'"
```

This should successfully work with some errors. On the first `git commit`, the PNG is changed by the hook and then readded to git. Therefore, the second `git commit` works. 

When running `git commit` with more than one file, the hook does not seem to finish. Run the following:

```shell
find src/icons/ -type f -name "*.png" -exec git add {} + && git commit -m 'add `src/icons` directory'
```

**Comment by redwarp** [comment]
So, I just tried, by just running `pre-commit run --all-files` and it's pretty much instant on my end :-/
So I can't really reproduce the issue?
All the files get modified, making the first time run fail. And the second succeed.
I'm using pre-commit version 2.20.0

**Comment by trueNAHO** [comment]
I followed the steps as specified in my first comment again after running `pre-commit clean` on my machine. The result was the same. I waited for 100min and it did not continue. 

Here some usefull output. I replaced `/home/$USER` with `~` in all the output messages. 

After running the first `find` command, I get:

```shell
$ find src/ -maxdepth 1 -type f -name "*.png" -print0 | xargs -0 -I {} bash -c "git add {} && git commit -m 'add {}'; git add {} && git commit -m 'add {}'"
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to ~/.cache/pre-commit/patch1662324761-33356.
Optimize PNG.............................................................Failed
- hook id: optimize-png
- files were modified by this hook
[INFO] Restored changes from ~/.cache/pre-commit/patch1662324761-33356.
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to ~/.cache/pre-commit/patch1662324762-33377.
Optimize PNG.............................................................Passed
[INFO] Restored changes from ~/.cache/pre-commit/patch1662324762-33377.
[master 9d4a1ae] add src/terminal_box_w.png
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 src/terminal_box_w.png
```

Note that this is not the full output. The rest of the output follows the same pattern.

After running the second `find` command and waiting for about 100min, I get:

```shell
$ find src/icons/ -type f -name "*.png" -exec git add {} + && git commit -m 'add `src/icons` directory'
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to ~/.cache/pre-commit/patch1662325095-34212.
Optimize PNG.............................................................
```

Then I tried the same on a different far more powerfull machine and the result was the same. Both machines run Arch Linux. I assume some package might be missing or something? But this is odd, since running `oxipng` directly works without any issue.

I installed `pre-commit` with `pacman -Syu python-pre-commit` and it currently runs on version 2.20.0.

Here some information about my main machine that might be usefull:

```shell
$ lscpu
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   36 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          2
On-line CPU(s) list:             0,1
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Celeron(R) CPU 887 @ 1.50GHz
CPU family:                      6
Model:                           42
Thread(s) per core:              1
Core(s) per socket:              2
Socket(s):                       1
Stepping:                        7
CPU(s) scaling MHz:              75%
CPU max MHz:                     1500.0000
CPU min MHz:                     800.0000
BogoMIPS:                        2994.82
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ht tm pbe syscall nx rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer xsave lahf_lm epb pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid xsaveopt dtherm arat pln pts md_clear flush_l1d
Virtualization:                  VT-x
L1d cache:                       64 KiB (2 instances)
L1i cache:                       64 KiB (2 instances)
L2 cache:                        512 KiB (2 instances)
L3 cache:                        2 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0,1
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Mitigation; PTE Inversion; VMX conditional cache flushes, SMT disabled
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT disabled
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Mmio stale data:   Unknown: No mitigations
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
```

```shell
$ sudo lshw -short
H/W path           Device    Class          Description
=======================================================
                             system         AO756 (AO756_0742_V2.04)
/0                           bus            Mimic
/0/0                         memory         128KiB BIOS
/0/4                         processor      Intel(R) Celeron(R) CPU 887 @ 1.50GHz
/0/4/9                       memory         32KiB L1 cache
/0/4/a                       memory         256KiB L2 cache
/0/4/b                       memory         2MiB L3 cache
/0/8                         memory         32KiB L1 cache
/0/18                        memory         4GiB System Memory
/0/18/0                      memory         DIMM [empty]
/0/18/1                      memory         4GiB SODIMM DDR3 Synchronous 1600 MHz (0.6 ns)
/0/100                       bridge         2nd Generation Core Processor Family DRAM Controller
/0/100/2                     display        2nd Generation Core Processor Family Integrated Graphics Controller
/0/100/16                    communication  7 Series/C216 Chipset Family MEI Controller #1
/0/100/1a                    bus            7 Series/C216 Chipset Family USB Enhanced Host Controller #2
/0/100/1a/1        usb1      bus            EHCI Host Controller
/0/100/1a/1/1                bus            Integrated Rate Matching Hub
/0/100/1a/1/1/1              communication  Bluetooth wireless interface
/0/100/1a/1/1/3              multimedia     HD WebCam
/0/100/1b                    multimedia     7 Series/C216 Chipset Family High Definition Audio Controller
/0/100/1c                    bridge         7 Series/C216 Chipset Family PCI Express Root Port 1
/0/100/1c.1                  bridge         7 Series/C210 Series Chipset Family PCI Express Root Port 2
/0/100/1c.1/0      wlp3s0    network        AR9462 Wireless Network Adapter
/0/100/1c.2                  bridge         7 Series/C210 Series Chipset Family PCI Express Root Port 3
/0/100/1c.2/0      enp4s0f0  network        NetLink BCM57785 Gigabit Ethernet PCIe
/0/100/1c.2/0.1    mmc0      bus            BCM57765/57785 SDXC/MMC Card Reader
/0/100/1d                    bus            7 Series/C216 Chipset Family USB Enhanced Host Controller #1
/0/100/1d/1        usb2      bus            EHCI Host Controller
/0/100/1d/1/1                bus            Integrated Rate Matching Hub
/0/100/1d/1/1/3              bus            USB2.0 Hub
/0/100/1d/1/1/3/1            input          USB Keyboard
/0/100/1d/1/1/3/2            input          USB Laser Mouse
/0/100/1f                    bridge         HM70 Express Chipset LPC Controller
/0/100/1f.2                  storage        7 Series Chipset Family 6-port SATA Controller [AHCI mode]
/0/100/1f.3                  bus            7 Series/C216 Chipset Family SMBus Controller
/0/1                         generic        PnP device MSF0001
/0/2                         generic        PnP device ETD0502
/0/3                         system         Motherboard registers
/0/5                         system         AT Real-Time Clock
/0/6                         generic        PnP device INT3f0d
/0/7                         system         Motherboard registers
/0/9                         system         System Board
```

The only thing I can do as of now is to remove this pre-commit hook from my repositories and manually run `oxipng`... :(

**Comment by redwarp** [comment]
So I'm currently not really able to try that easily, because I don't have any arch linux installed, and I'm just on a laptop with windows running right now.
I have a question though: Do you have the same issue with the version 1.1.0 of the pre-commit hook?
In the 1.2.0, I added some parallelization library to the code, and I'm thinking that there might be a threading problem happening for some reason.
It would be interesting to know if you have the same issue with the 1.1.0.

**Comment by trueNAHO** [comment]
Nice thinking there! Version 1.1.0 worked and it only took a second!

Now a more detailed description. The first `find` command worked same as before. The second `find` command worked correctly this time. Here is the ouptut of the second `find` commmand, which only took one second to execute (replacing `/home/$USER` with `~`):

```shell
$ find src/icons/ -type f -name "*.png" -exec git add {} + && git commit -m 'add `src/icons` directory'
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to ~/.cache/pre-commit/patch1662381234-7324.
Optimize PNG.............................................................Failed
- hook id: optimize-png
- files were modified by this hook
[INFO] Restored changes from ~/.cache/pre-commit/patch1662381234-7324.
```

After running the same command again, which also only took one second to execute, I get:

```shell
$ find src/icons/ -type f -name "*.png" -exec git add {} + && git commit -m 'add `src/icons` directory'
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to ~/.cache/pre-commit/patch1662381256-7359.
Optimize PNG.............................................................Passed
[INFO] Restored changes from ~/.cache/pre-commit/patch1662381256-7359.
[master 64039d9] add `src/icons` directory
 33 files changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 src/icons/antergos.png
 create mode 100644 src/icons/arch.png
 create mode 100644 src/icons/arcolinux.png
 create mode 100644 src/icons/arcolinuxb.png
 create mode 100644 src/icons/arcolinuxd.png
 create mode 100644 src/icons/chakra.png
 create mode 100644 src/icons/crunchbang.png
 create mode 100644 src/icons/debian.png
 create mode 100644 src/icons/deepin.png
 create mode 100644 src/icons/edubuntu.png
 create mode 100644 src/icons/elementary.png
 create mode 100644 src/icons/fedora.png
 create mode 100644 src/icons/freebsd.png
 create mode 100644 src/icons/frugalware.png
 create mode 100644 src/icons/gentoo.png
 create mode 100644 src/icons/gnu-linux.png
 create mode 100644 src/icons/kali.png
 create mode 100644 src/icons/kaos.png
 create mode 100644 src/icons/korora.png
 create mode 100644 src/icons/kubuntu.png
 create mode 100644 src/icons/lfs.png
 create mode 100644 src/icons/linux-mint.png
 create mode 100644 src/icons/lubuntu.png
 create mode 100644 src/icons/mageia.png
 create mode 100644 src/icons/manjaro.png
 create mode 100644 src/icons/opensuse.png
 create mode 100644 src/icons/parrot.png
 create mode 100644 src/icons/siduction.png
 create mode 100644 src/icons/slackware.png
 create mode 100644 src/icons/steamos.png
 create mode 100644 src/icons/ubuntu.png
 create mode 100644 src/icons/windows.png
 create mode 100644 src/icons/xubuntu.png
```

It seems like version 1.2.0 is the problem.

**Comment by redwarp** [comment]
Hum, then I'll probably revert and remove these change then! I'm not sure the parallelization is that useful, and it obviously creates problems!
On top of that, I believe that the latest version of pre-commit can itself run multipler optimizer in parallel.
Glad it worked for you!

**Comment by trueNAHO** [comment]
Thanks for the support! We should close this issue once the new "revert" version has been released. Probably something like v1.2.1.

**Comment by redwarp** [comment]
Yup!

**Comment by redwarp** [comment]
Done!
FYI, it seems that a hook was added directly in the oxipng repo (if I read https://pre-commit.com/hooks.html properly)
I haven't tried it, but it might make this hook redundant?

**Comment by redwarp** [comment]
(Tagged 1.2.1)

**Comment by trueNAHO** [comment]
The [latest oxipng tag (v5.0.1)](https://github.com/shssoichiro/oxipng/releases/tag/v5.0.1) does not seem to include the `.pre-commit-hooks.yaml` file yet. To use this hook nonetheless, you would have to manually add this missing file to the `pre-commit` cache. At least that is what I had to do. Until [oxipng](https://github.com/shssoichiro/oxipng) releases a newer tag, it seems like this hook is not redundant at all.

**Comment by trueNAHO** [comment]
As a note for anyone reading this issue. A new [oxipng](https://github.com/shssoichiro/oxipng) tag has been released: [v6.0.0](https://github.com/shssoichiro/oxipng/releases/tag/v6.0.0). It seems to include the `.pre-commit-hooks.yaml` file now.

**Comment by redwarp** [comment]
I tested the pre-commit from oxipng, it seems to work well!
Probably time for me to deprecate this one ;-)

**Comment by redwarp** [comment]
I released one last version, 1.2.2, using the latest 6.0.1 oxipng, but will also print a deprecation message telling people "use real oxipng instead"
Thanks for letting me know!
